symbolic-base/src/ZkFold/Symbolic/Data/JWT.hs

```diff
@@ -5,123 +5,73 @@
+import           GHC.TypeLits                       (Symbol)
+import           Prelude                            (type (~), ($), (.))
 import qualified Prelude                            as P
-import           Test.QuickCheck                    (Arbitrary (..))
 
-import           ZkFold.Base.Algebra.Basic.Class
 import           ZkFold.Base.Algebra.Basic.Number
-import           ZkFold.Base.Data.HFunctor          (hmap)
 import qualified ZkFold.Base.Data.Vector            as V
-import           ZkFold.Base.Data.Vector            ((!!))
-import           ZkFold.Symbolic.Algorithms.RSA
 import           ZkFold.Symbolic.Class
 import           ZkFold.Symbolic.Data.Bool
-import           ZkFold.Symbolic.Data.ByteString    (ByteString (..), concat, toWords)
 import           ZkFold.Symbolic.Data.Class
-import           ZkFold.Symbolic.Data.Combinators
-import           ZkFold.Symbolic.Data.Eq
-import           ZkFold.Symbolic.Data.FieldElement
+import           ZkFold.Symbolic.Data.Combinators   hiding (toBits)
 import           ZkFold.Symbolic.Data.Input         (SymbolicInput)
-import           ZkFold.Symbolic.Data.UInt
+import           ZkFold.Symbolic.Data.JWT.Utils
 import qualified ZkFold.Symbolic.Data.VarByteString as VB
-import           ZkFold.Symbolic.Data.VarByteString (VarByteString (..), wipeUnassigned, (@+))
-import           ZkFold.Symbolic.MonadCircuit       (newAssigned)
-
+import           ZkFold.Symbolic.Data.VarByteString (VarByteString (..), (@+))
 
+-- | Types than can be represented as a Symbolic JSON string
+--
 class IsSymbolicJSON a where
     type MaxLength a :: Natural
 
     toJsonBits :: a -> VarByteString (MaxLength a) (Context a)
 
-
--- | RSA Public key with Key ID
+-- | Types than can be serialised
 --
-data Certificate ctx
-    = Certificate
-        { pubKid :: VarByteString 320 ctx
-        , pubKey :: PublicKey 2048 ctx
-        }
-    deriving Generic
+class IsBits a where
+    type BitCount a :: Natural
 
-deriving instance
-    ( P.Eq (PublicKey 2048 ctx)
-    , P.Eq (VarByteString 320 ctx)
-    ) => P.Eq (Certificate ctx)
-deriving instance
-    ( P.Show (PublicKey 2048 ctx)
-    , P.Show (VarByteString 320 ctx)
-    ) => P.Show (Certificate ctx)
-deriving instance
-    ( NFData (PublicKey 2048 ctx)
-    , NFData (VarByteString 320 ctx)
-    ) => NFData (Certificate ctx)
-instance
-    ( SymbolicData (PublicKey 2048 ctx)
-    , Symbolic ctx
-    ) => SymbolicData (Certificate ctx)
-instance
-    ( SymbolicInput (PublicKey 2048 ctx)
-    , Symbolic ctx
-    ) => SymbolicInput (Certificate ctx)
+    toBits :: a -> VarByteString (BitCount a) (Context a)
 
-
--- | RSA Private key with Key ID
+-- | Signing algorithm for JWT (such as RS256)
 --
-data SigningKey ctx
-    = SigningKey
-        { prvKid :: VarByteString 320 ctx
-        , prvKey :: PrivateKey 2048 ctx
-        }
-    deriving Generic
+class SigningAlgorithm (alg :: Symbol) where
```

Comment:
So in essence, a `SigningAlgorithm` is a dict of types. With `DataKinds` and `TypeFamilies`, you can declare it as such:
```haskell
data SigningAlgorithm = MkSigningAlgorithm
  { saSKey :: Type
  , saVKey :: Type
  , saSignature :: Type
  , saHash :: Type
  }

type family SKey (_ :: SigningAlgorithm) :: Type where SKey (MkSigningAlgorithm skey _ _ _) = skey
...

type RS256 ctx = MkSigningAlgorithm (SigningKey ctx) (Certificate ctx) (ByteString 2048 ctx) (ByteString 256 ctx)
```

Reply:
What advantage will it provide? From the user's perspective, the only thing which will change will be calling `signPayload @(RS256 ctx)` instead of `signPayload @"RS256"`, right?

Reply:
Yes, that's right
Also this would allow to create new signing algorithms inplace (though I don't know if it's useful)
It's just that I'm not in favor of introducing typeclasses where some other mechanism would suffice, but you're right that here the difference is really minor